### PR TITLE
Check AWS Account ID (#1661)

### PR DESCRIPTION
**Pull request by amarjandu**
Verifies that an operators AWS Account ID matches the the branch/deployment Account ID.

Connected To: #1661

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
By our own convention, a PR author is implicitly assigned. So no one assigned means author is assigned.

File: README.md [hunk 1]
```diff
@@ -545,6 +545,8 @@ make delete
 
 # 4. Running indexer or service locally
 
+**Ensure infrastructure for lambda(s) are deployed before trying to run locally. See Sections: [2](#2-getting-started), [3](#3-deployment)**
```
Comment (hannes-ucsc) [inline README.md:548]:
```suggestion
While this method *does* run the service or indexer locally on your machine, it still requires that the cloud resources used by them are already deployed. See sections [2](#2-getting-started) and [3](#3-deployment) on how to do that.
```

and check if I botched the links. And wrap at 80.

File: common.mk [hunk 2]
```diff
@@ -33,6 +33,16 @@ ifneq ($(shell python -c "import wheel as w; \
 $(error Looks like the `wheel` package is outdated or missing. See README for instructions on how to fix this.)
 endif
 
+# This check should not occur within CI environments, where AWS Credentials might not be supplied
+ifneq ($(shell python -c "import os, sys; \
+                          import boto3 as b; \
+                          print(True) if os.getenv('CI') else \
+                          print(b.client('sts').get_caller_identity()['Account'] == \
+                          os.environ['AZUL_AWS_ACCOUNT_ID'])"),True)
```
Comment (hannes-ucsc) [inline common.mk:41]:
```suggestion
ifneq ($(shell python -c "import os, sys; \
                          import boto3 as b; \
                          print(os.getenv('CI') or \
                          b.client('sts').get_caller_identity()['Account'] == os.environ['AZUL_AWS_ACCOUNT_ID'])"),True)
```

and don't mix os.getenv with os.environ

File: common.mk [hunk 3]
```diff
@@ -33,6 +33,16 @@ ifneq ($(shell python -c "import wheel as w; \
 $(error Looks like the `wheel` package is outdated or missing. See README for instructions on how to fix this.)
 endif
 
+# This check should not occur within CI environments, where AWS Credentials might not be supplied
+ifneq ($(shell python -c "import os, sys; \
+                          import boto3 as b; \
+                          print(True) if os.getenv('CI') else \
```
Comment (hannes-ucsc) [inline common.mk:39]: Another thing: check if Gitlab defines that variable. If not, this might fail on our Gitlab instance.

Comment (amarjandu) [reply common.mk:39]: It's used within both:
Travis: https://docs.travis-ci.com/user/environment-variables/#default-environment-variables
Gitlab: https://docs.gitlab.com/ee/ci/variables/predefined_variables.html

Was there any other CI that Azul is tested with?

Comment (hannes-ucsc) [reply common.mk:39]: Just those two.

File: common.mk (hunk 2 above)
Comment (amarjandu) [reply common.mk:41]: @hannes-ucsc i'll switch to getenv for both.

Comment (hannes-ucsc) [reply common.mk:41]: I should have been more clear. There is precedence in common.mk and the rest of Azul for `os.environ` and you should use that. The reason we prefer `os.environ` is that we can more clearly express whether we do or do not expect a variable to be present(`os.environ['…']` vs `os.environ.get('…')`).

Comment (amarjandu) [reply common.mk:41]: in this case i dont expect an operators machine to have the env var `CI` set.
In that case if we were to use `os.environ['CI']` there would be KeyError raised when an operator would run any Make command. 

I could perform additional checks within the env files that could set CI to False or True,and move to explicitly check if CI is set to True. However there would be additional complexity added for both Travis and Gitlab. 

Unless there is an easier way. 🤔

File: README.md (hunk 1 above)
Comment (amarjandu) [reply README.md]: Done.

File: common.mk (hunk 2 above)
Comment (hannes-ucsc) [reply common.mk]: My post on Slack should resolve this. Let me know if it does not.

Comment (amarjandu) [reply common.mk]: @hannes-ucsc  There was a new issue with using the `or` statement within the print, Travis and Gitlab both set the CI value to `true` while our comparison is for a boolean `True`; it ends up always = failin due to `true != True`. 
Was considering using https://docs.python.org/3/distutils/apiref.html#distutils.util.strtobool to get around this.

Comment (amarjandu) [reply common.mk]: `strtobool` fails to handle None case, considering; 
```
('true' if b.client('sts').get_caller_identity()['Account'] == os.environ['AZUL_AWS_ACCOUNT_ID']))\
```
and setting the comparison to `true` within the make statement. 

```
ifneq ($(shell python -c "import os, sys; \
                          import boto3 as b; \
                          print(os.environ.get('CI') or \
                          ('true' if b.client('sts').get_caller_identity()['Account'] == os.environ['AZUL_AWS_ACCOUNT_ID']))\
                          "),true)
```

Comment (hannes-ucsc) [reply common.mk]: I'll reply on the most recent changes.

File: common.mk [hunk 4]
```diff
@@ -33,6 +33,16 @@ ifneq ($(shell python -c "import wheel as w; \
 $(error Looks like the `wheel` package is outdated or missing. See README for instructions on how to fix this.)
 endif
 
+# This check should not occur within CI environments, where AWS Credentials might not be supplied
+ifneq ($(shell python -c "import os, sys; \
+                          import boto3 as b; \
+                          print(os.environ.get('CI') or \
```
Comment (hannes-ucsc) [inline common.mk:39]:
```suggestion
                          print(os.environ.get('CI') == 'true' or \
```

Comment (amarjandu) [reply common.mk]: 👍

**Comment by amarjandu** [comment]
@hannes-ucsc regarding the addition of `azul.config.aws_account` I added the functionality in the commit that resolves the issue, but added a fixup to remove it  after Jesse's [request](https://github.com/DataBiosphere/azul/pull/1664#pullrequestreview-385890685).

Although I added the functionality to the Config class it seemed a little out-of-place when using it. 
for example with terraform templates
```
f"arn:aws:sns:{aws.region_name}:{config.aws_account}:cloudwatch-alarms",
f"arn:aws:sns:{aws.region_name}:{config.aws_account}:dcp-events"
```

while for something like `scripts/check_branch.py` the modification "fits" easier.
```
    stage_by_branch = config.main_deployments_by_branch
    account = azul.config.aws_account
```

The idea that they are "out-of-place" stems from there being 2 classes where the information can come. Within the TF template, the use of the `config` class to get the `aws_account` seems strange, especially because the `aws` class can give that information without additional imports. 

I can re-add the functionality, but thought it should be brought up, if its a non-issue, please let me know. :)

**Comment by hannes-ucsc** [comment]
I wouldn't interpret @jessebrennan post as a request to remove all the changes to the `azul.config` and `azul.deployment.aws` classes. Let's go with what you and I agreed to and we'll take it from there in my next review.

File: common.mk [hunk 5]
```diff
@@ -33,6 +33,16 @@ ifneq ($(shell python -c "import wheel as w; \
 $(error Looks like the `wheel` package is outdated or missing. See README for instructions on how to fix this.)
 endif
 
+# This check should not occur within CI environments, where AWS Credentials might not be supplied
+ifneq ($(shell python -c "import os, sys; \
+                          import boto3 as b; \
+                          print(os.environ.get('CI') == 'true' or \
+                          b.client('sts').get_caller_identity()['Account'] == os.environ['AZUL_AWS_ACCOUNT_ID'])\
+                          "),True)
+$(error Looks like there is a mismatch between the AWS account you have configured, and what the deployment expects. \
```
Comment (hannes-ucsc) [inline common.mk:42]:
```suggestion
$(error Looks like there is a mismatch between AZUL_AWS_ACCOUNT_ID and the currently active AWS credentials. \
```

File: common.mk [hunk 6]
```diff
@@ -33,6 +33,16 @@ ifneq ($(shell python -c "import wheel as w; \
...
+        Compare output from 'aws sts get-caller-identity' and the deployment environment file)
```
Comment (hannes-ucsc) [inline common.mk:43]:
```suggestion
        Check the output from 'aws sts get-caller-identity' against the value of that environment variable.)
```

**Comment by amarjandu** [comment]
@hannes-ucsc our proposed solution to asserting that the dynamically evaluated AccountID is equivalent to the static AccountID from the Environment Variable does not seem to past out CI testing environment, I'm not sure that we provide Travis with AWS Credentials, as i dont see any within the ENV configuration, however during run time on Travis, the dynamic account value is [123456789012](https://travis-ci.org/github/DataBiosphere/azul/builds/670764112#L697).

**Comment by hannes-ucsc** [comment]
> I'm not sure that we provide Travis with AWS Credentials

We do not and that's intentional.

> This causes a break on the assertion check that we have placed.

I am surprised that that code (`azul.deployment.aws.account`) is even executed on Travis. Since Travis does not have AWS creds, that property would have to raise an exception on `develop`. Why does it not?

Another thing: Gitlab does have cloud access and we don't need to circumvent the common.mk check there. AFAICT, the current code circumvents the check on both GL and Travis. It only needs to be circumvented on Travis.

**Comment by amarjandu** [comment]
> Another thing: Gitlab does have cloud access and we don't need to circumvent the common.mk check there. AFAICT, the current code circumvents the check on both GL and Travis. It only needs to be circumvented on Travis.

I've updated the `common.mk` to ensure that the `make` checks are only bypassed within the `TRAVIS` CI system.

I've also adjusted the `azul_test_case` class to modify the environment value allowing `unit_test` to pass like we discussed after standup.

File: attic/deployments/dev/environment [hunk 7]
```diff
@@ -30,6 +30,8 @@ _set AZUL_DEBUG 1
 
 _set AZUL_OWNER hannes@ucsc.edu
 
+_set AZUL_AWS_ACCOUNT_ID 122796619775
```
Comment (hannes-ucsc) [inline attic/deployments/dev/environment:33]: The attic is for "deleted" files. We don't need to, and should not, modify them.

File: attic/deployments/sandbox/environment [hunk 8]
```diff
@@ -39,6 +39,8 @@ _set AZUL_DEBUG 1
 
 _set AZUL_OWNER hannes@ucsc.edu
 
+_set AZUL_AWS_ACCOUNT_ID 122796619775
```
Comment (hannes-ucsc) [inline attic/deployments/sandbox/environment:42]: Same here

File: test/azul_test_case.py [hunk 9]
```diff
@@ -56,6 +59,9 @@ class AzulTestCase(TestCase):
     def setUpClass(cls) -> None:
         super().setUpClass()
 
+        cls.aws_account_id = config.aws_account_id
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:62]:
```suggestion
        cls.aws_account_id = os.environ['AZUL_AWS_ACCOUNT_ID']
```

For symmetry. There is no point in using the abstraction if the line below does not. The long term solution would be to extend the abstraction to handle patches like this.

File: test/azul_test_case.py [hunk 10]
```diff
@@ -56,6 +59,9 @@ class AzulTestCase(TestCase):
...
+        os.environ['AZUL_AWS_ACCOUNT_ID'] = '123456789012'
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:63]: You should mention in a comment where the value comes from or, even better, use the corresponding moto constant here.

File: test/azul_test_case.py [hunk 11]
```diff
@@ -1,5 +1,7 @@
+import os
 from unittest import TestCase
 
+from moto.organizations import utils
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:4]:
```suggestion
from moto.organizations.utils import MASTER_ACCOUNT_ID as moto_account_id
```

File: test/azul_test_case.py [hunk 12]
```diff
@@ -56,6 +58,12 @@ class AzulTestCase(TestCase):
     def setUpClass(cls) -> None:
         super().setUpClass()
 
+        # Set the AZUL_AWS_ACCOUNT_ID to what the moto package is using, this ensures
+        # that there will not be assertion errors when using the azul.deployment.aws.account
+        # property.
+        cls.aws_account_id = os.environ['AZUL_AWS_ACCOUNT_ID']
+        os.environ['AZUL_AWS_ACCOUNT_ID'] = utils.MASTER_ACCOUNT_ID
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:65]:
```suggestion
        os.environ['AZUL_AWS_ACCOUNT_ID'] = moto_account_id
```

is more clear IMO

**Comment by hannes-ucsc** [comment]
Despite the scope creep, I think the estimate of 3 is still accurate.

**Comment by amarjandu** [comment]
> https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/1647

ah, I should have checked for usage of the object; I've repaired the error, awaiting testing .

**Comment by hannes-ucsc** [comment]
Your changes in response to my last review should have been made as a separate !fixup commit.

**Comment by amarjandu** [comment]
> Your changes in response to my last review should have been made as a separate !fixup commit.

@hannes-ucsc 
Strange IIRC that was what i had done. Ill keep my eyes on that, thanks for informing me. 

Screen Shot 2020-04-16 at 10 59 55 AM

**Comment by hannes-ucsc** [comment]
> Strange IIRC that was what i had done. Ill keep my eyes on that, thanks for informing me.

I think you are right. My force push actually shows that I obliterated the fixup commit in your screenshot.

image

And https://github.com/DataBiosphere/azul/commit/4cb2db59c4958e7d9014db5e376368c4ec50b08b only contains the changes that address my review comments.

My apologies.

**Comment by hannes-ucsc** [comment]
The IT fail seems to be limited to this branch. A subsequent sandbox build on a different branch succeeded: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/1760

**Comment by hannes-ucsc** [comment]
This was merged.

**Pull request closed without merging**